Fix `topk` kernel for 1D non-contiguous output tensors.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62942 Fix `topk` kernel for 1D non-contiguous output tensors.**
* #62941 Fix `sort` kernel for 1D non-contiguous output tensors.

Fixes: #62940

Apparently, `topk` had the same problems as `sort`: ignoring the strides of 1D tensors.

ysiraichi committed:
Fix `topk` kernel for 1D non-contiguous output tensors.

Fixes: #62940

Apparently, `topk` had the same problems as `sort`: ignoring the strides of 1D tensors.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/TensorTopK.cu:
```diff
@@ -234,9 +234,14 @@ TORCH_IMPL_FUNC(topk_out_cuda)
+    int collapseTopKDim = 0;                                              \
+    int collapseIndicesDim = 0;                                           \
+    if (input.dim() > 1) {                                                \
+      int collapseInputDim = inputInfo.collapseDims(dim);                 \
```
this doesn't look right, these values are limited to `if` scope.

ysiraichi replied:
Ah. You are completely right! Thanks.

ysiraichi committed:
Update on "Fix `topk` kernel for 1D non-contiguous output tensors."


Fixes: #62940

Apparently, `topk` had the same problems as `sort`: ignoring the strides of 1D tensors.

[ghstack-poisoned]

heitorschueroff commented:
@ysiraichi Thank you for your help in fixing this problem. The fix was a bit more intricate then it seemed so @ngimel submitted a PR here https://github.com/pytorch/pytorch/pull/63029.
